UnitControl: Update unit dropdown design

mirka opened the pull request:
Part of #41973

## What?

Updates the unit dropdown to the new designs for the 40px size variant.

## Why?

To get our components closer to the mockups.

## How?

Add conditional styles for the dropdown depending on the size variant.

## Testing Instructions

1. `npm run storybook:dev`
2. See the stories for UnitControl.

### Known issues

- #42650

## Screenshots or screencast <!-- if applicable -->

https://user-images.githubusercontent.com/555336/181043192-35466198-5e32-47b7-970a-ab559d47a1f6.mp4

### When the unit isn't a dropdown (i.e. single unit case)

<img src="https://user-images.githubusercontent.com/555336/176149227-6305ea21-78d2-4e98-8580-14ac16afc992.png" alt="Dropdown text has no accent color when the dropdown is inactive" width="120">

mirka committed:
Get `size` type from ancestor

mirka committed:
Remove dropdown color when single unit

mirka commented:
@pablohoneyhoney @jameskoster  Hi 👋 Can we get some guidance on the interaction details (touch target boundary, hover/focus styles) of this unit dropdown? For reference, these are the hover/focus styles we had for the previous design:

<img src="https://user-images.githubusercontent.com/555336/176152413-d2351a14-a146-4982-8de9-6c844e9eaf80.png" alt="Hover style" width="120"> <img src="https://user-images.githubusercontent.com/555336/176152418-9dfa474f-595f-4ff7-a91c-a58becdc7c06.png" alt="Focus style" width="120">

Similarly, we'll soon need guidance for the new step button designs in the NumberControl as well. (Not in this PR though)

<img width="132" alt="image" src="https://user-images.githubusercontent.com/555336/176152874-4161e428-6a6e-4cb0-9935-fa40e03241c5.png">

jameskoster commented:
Assuming the following anatomy:

<img width="690" alt="Screenshot 2022-06-28 at 12 11 49" src="https://user-images.githubusercontent.com/846565/176165390-0798a8aa-6e62-4c91-84f9-740b58274458.png">

Where state styling is informed by existing components (`InputControl` and `Button`), we'd end up with the following:

<img width="593" alt="Screenshot 2022-06-28 at 12 14 18" src="https://user-images.githubusercontent.com/846565/176165558-b41dd91b-5bfb-4c0a-98a5-0d95ae8050d1.png">

Here's a gif demonstrating the interactions:

<img src="https://user-images.githubusercontent.com/846565/176167351-4a67aa5f-e10a-4236-8aeb-3329792e900f.gif" width="400">

To me this seems like a reasonable place to start that avoids changing other components (better handled separately). But I'll defer to Pablo on how to proceed.

mirka commented on packages/components/src/unit-control/styles/unit-control-styles.ts:
```diff
@@ -96,6 +119,7 @@ export const UnitSelect = styled.select< SelectProps >`
 		cursor: pointer;
 		border: 1px solid transparent;
 		height: 100%;
+		font-family: inherit;
```
Similar to #38969, this was incorrectly rendering in Arial (per Chrome user agent style).

mirka commented on packages/components/src/unit-control/types.ts:
```diff
@@ -54,12 +54,6 @@ export type UnitSelectControlProps = {
 	 * A callback function invoked when the value is changed.
 	 */
 	onChange?: UnitControlOnChangeCallback;
-	/**
-	 * Size of the control option. Supports "default" and "small".
-	 *
-	 * @default 'default'
-	 */
-	size?: SelectSize;
```
Moved to a direct import from `InputControlProps` so it doesn't get out of date.

jameskoster commented:
An update to the concept above:

It looks like I was using a slightly outdated Figma component with the wrong button size. We have two options for the unit button:

1. Use the Tertiary button variant (36px height)
2. Use the Small button size (24px height)

### Tertiary
<img src="https://user-images.githubusercontent.com/846565/176408786-c1ad2781-7aaa-4c2f-988c-e61aee94d0ab.gif" width="400">

### Small
<img src="https://user-images.githubusercontent.com/846565/176409833-65ee9a3c-2b02-407c-bf87-2d38ef03579d.gif" width="400">

Thinking ahead to the Stepper component mentioned above, I think it makes sense to use the Tertiary variant because its dimensions match the Icon button:

<img width="423" alt="Screenshot 2022-06-29 at 11 06 29" src="https://user-images.githubusercontent.com/846565/176410938-1f24c0a0-14ed-4bb6-a620-280f69dde35d.png">

The only small detail we'll need to consider is the color difference:

<img width="417" alt="Screenshot 2022-06-29 at 11 08 57" src="https://user-images.githubusercontent.com/846565/176411485-39882275-80cf-45e1-8325-ebb45ae23ac4.png">

mirka committed:
Remove unnecessary control type config

mirka committed:
UnitControl: Update unit dropdown design

mirka committed:
Fixup eslint issue

mirka committed:
Update large unit styles

mirka commented:
@jameskoster Now for the final touches ✨ Here it is with the exact same hover/focus styles as Tertiary Button:

https://user-images.githubusercontent.com/555336/180580953-90c5a677-8182-4f91-a293-17f81d2e15d1.mp4

(Tertiary button has a 1px inset hover shadow, and a 1.5px outset focus shadow.)

It's a bit cramped. I'll note that, at least from a technical library maintenance standpoint, we don't necessarily have to follow the Button component sizes/styles — we can easily make a new component specifically to use inside text fields.

There are a few ways we can tweak the text field button, like:

- Make it slightly smaller than 36px.
- Change the hover/focus styles.

Feel free to write down the styles you want, or push the changes directly to this branch (I can take care of any final refactoring that may be necessary).

---

## Known issues

- The `text-align` bug in Safari (#42650) is more noticeable in this new design 🥲

jameskoster commented:
Hmm yeah that is a bit tight isn't it. Especially hovering the unit button after focussing the input.

Looking at the 'Remove color' button from global styles (which is somewhat similar in terms of general UI) we can probably try `small` `icon` Buttons:

<img width="321" alt="Screenshot 2022-07-25 at 10 37 26" src="https://user-images.githubusercontent.com/846565/180747054-6de072fd-7b42-4df6-9f1a-c4a5c7f6bc41.png">

Sorry for flip-flopping 😓

mirka committed:
Decrease size to 24px

mirka commented:
> Sorry for flip-flopping 😓

No problem at all 😄

Any preferences on how to deal with overflows? A 24px square is pretty small and will already overflow with common units like `em`/`rem`.

<img width="116" alt="Overflowing with 'em' unit" src="https://user-images.githubusercontent.com/555336/180880155-3fb0de6b-b197-469f-bda0-207db4d6abaa.png">

We can for example set some kind of `max-width`, and grow to the longest option label within that max.

https://user-images.githubusercontent.com/555336/180882081-8e70c074-1d18-4a9b-9473-5568d1d17ecb.mp4

This means that most UnitControls will have rectangular dropdown buttons, and the proportions of those rectangles will vary depending on the longest available option in that UnitControl. That subtle variance across UnitControls may not be desirable. In that case it might be better to just set all dropdown widths to the maximum width regardless of the option lengths.

ciampo commented:
> In that case it might be better to just set all dropdown widths to the maximum width regardless of the option lengths.

That would be the best choice, IMO (plus maybe a conservative, safety `max-width` in case the unit label is extremely long?)

jameskoster commented:
Oh, I think it would be ok for the button to hug the current value, probably with a 24px min-width, and a max-width as @ciampo suggested.

mirka committed:
Remove max width from stories

We need to know whether the width is unconstrained by default, and we need to be able to test `__unstableInputWidth`.

mirka committed:
Apply overflow handling

mirka commented on packages/components/src/unit-control/stories/index.tsx:
```diff
@@ -24,9 +24,6 @@ const meta: ComponentMeta< typeof UnitControl > = {
 		__unstableStateReducer: {
 			control: { type: null },
 		},
-		size: {
-			control: { type: 'select' },
-		},
```
Removed this explicit config because Storybook will use a `radio` by default.

mirka commented on packages/components/src/unit-control/stories/index.tsx:
```diff
@@ -54,16 +51,14 @@ const DefaultTemplate: ComponentStory< typeof UnitControl > = ( {
 	const [ value, setValue ] = useState< string | undefined >( '10px' );
 
 	return (
-		<div style={ { maxWidth: '100px' } }>
```
Removed these max-width containers in the stories because we need to know whether the width is unconstrained by default, and we need to be able to test `__unstableInputWidth`.

mirka commented:
Thanks y'all, I've updated the OP with a screen recording of the latest styles. This PR is now ready for final review.

The Safari issue (#42650) is pretty annoying. I don't have the appetite to address it in this PR, but we might eventually need to move to a custom select implementation if we want the centered design in Safari. It's unlikely that the bug will be fixed in Safari itself — that bug report just celebrated its 12th birthday 😇

ciampo commented on packages/components/src/unit-control/styles/unit-control-styles.ts:
```diff
@@ -59,62 +60,107 @@ export const ValueInput = styled( NumberControl )`
+		`,
+		large: css`
+			box-sizing: border-box;
+			min-width: 24px;
```
I still don't have a clear idea of when we should use the `space` utility and when we should just write "raw" px values. What do you think?

mirka replied:
😂 

I think my current rule of thumb is:

- If it's less than 4px, use the exact value
- If it's size variant value, use the exact value (for readability, and because there's a 30px variant!)
- If it's related to spacing (margin/padding), prefer to use `space`

Assuming the main point of `space` is to encourage people to use multiples of 4px, it shouldn't _really_ matter whether the `space` util is used 🤞

ciampo replied:
Sounds reasonable

ciampo commented on packages/components/src/unit-control/styles/unit-control-styles.ts:
```diff
@@ -59,62 +60,107 @@ export const ValueInput = styled( NumberControl )`
+			margin-inline-end: ${ space( 2 ) };
+			padding: ${ space( 1 ) };
+			color: ${ COLORS.ui.theme };
+			font-size: 13px;
```
Thought: I'd love to start standardizing font-sizes across components and be able to provide a "typography palette" through the various components from the library (e.g `Heading`, `Text`..)

ciampo commented on packages/components/src/unit-control/styles/unit-control-styles.ts:
```diff
@@ -59,62 +60,107 @@ export const ValueInput = styled( NumberControl )`
+				color: ${ COLORS.ui.borderFocus };
+				box-shadow: inset 0 0 0
+					${ CONFIG.borderWidth + ' ' + COLORS.ui.borderFocus };
+				outline: ${ CONFIG.borderWidth } solid transparent; // For High Contrast Mode
```
Do we need to test the look of the component in "high contrast mode" in any particular way?

mirka replied:
As far as I know, there isn't a way to test it outside of a Windows environment. The transparent outline thing should be pretty straightforward though.

ciampo reviewed (commented):
Code changes LGTM (I just left a few minor comments) 

https://user-images.githubusercontent.com/1083581/181062442-c98eb172-06fe-4f7d-8e57-c5ad4bd04f19.mp4

I'll leave it up to @jameskoster for the final approval

jameskoster commented:
Is it possible to render the button at a width that simply hugs the selected value? At the moment width seems to be based on the longest label in the list of `option`s, which can have some awkward manifestations when there's a mixture of short/long labels:

<img width="258" alt="Screenshot 2022-07-27 at 12 49 36" src="https://user-images.githubusercontent.com/846565/181239846-67be3c34-7356-4764-ba4d-d9855ba46d9e.png">

Something like this would be a bit better:

<img width="239" alt="Screenshot 2022-07-27 at 12 51 35" src="https://user-images.githubusercontent.com/846565/181240123-c1fd3c71-e246-4f32-a3c2-d77c0ba812dc.png">

That would enable us to have a greater max-width value as well which I think could be valuable. 48px is a bit tight.

mirka commented:
> Is it possible to render the button at a width that simply hugs the selected value?

It will be possible if we move to a [custom select implementation](https://github.com/WordPress/gutenberg/pull/42000#issuecomment-1195642180), which will allow us to fix the Safari issue as well. My recommendation is to merge for now with the current implementation, and move that to a follow-up issue since the task scope will be a bit big. (Unless you feel it's a blocker.)

> That would enable us to have a greater max-width value as well which I think could be valuable. 48px is a bit tight.

Yes I was thinking about this too, and it will take a bit of fiddling to get the logic right. The complexity is in maintaining a reasonable flex width balance between the prefix, number, and suffix elements. For example, in a tight context like the Border control:

<img width="268" alt="A BorderControl field that has a prefix, number field, and unit dropdown all in a 110px width" src="https://user-images.githubusercontent.com/555336/181244681-40c6ba1a-4b79-44de-b61b-e81c144016b8.png">

It might even be that we'll need to allow the consumer to customize the width via props, so it makes sense in different contexts.

mirka commented:
📌 I'm going to hold off on merging this while I look into #42717 first. The unit dropdown looks fine in Storybook but completely breaks down in WP 😭

ciampo commented:
> Yes I was thinking about this too, and it will take a bit of fiddling to get the logic right. The complexity is in maintaining a reasonable flex width balance between the prefix, number, and suffix elements.

Yeah, I can totally see that. I would personally try to leverage flexbox and model an elastic behavior

> 📌 I'm going to hold off on merging this while I look into https://github.com/WordPress/gutenberg/issues/42717 first. The unit dropdown looks fine in Storybook but completely breaks down in WP 😭

Good catch

jameskoster commented:
> Unless you feel it's a blocker.

I don't think so :)

mirka committed:
Add changelog

mirka committed:
Add back handling for forms.css

mirka commented:
Ok, I've fixed the #42717 problem, the unit dropdown now looks the same in isolated Storybook and WP 👍

This is now ready for final review.

ciampo reviewed (approved):
LGTM 🚀 

We can look into refining component in follow-ups:

> > Is it possible to render the button at a width that simply hugs the selected value?
> 
> It will be possible if we move to a [custom select implementation](https://github.com/WordPress/gutenberg/pull/42000#issuecomment-1195642180), which will allow us to fix the Safari issue as well. My recommendation is to merge for now with the current implementation, and move that to a follow-up issue since the task scope will be a bit big. (Unless you feel it's a blocker.)
> 
> > That would enable us to have a greater max-width value as well which I think could be valuable. 48px is a bit tight.
> 
> Yes I was thinking about this too, and it will take a bit of fiddling to get the logic right. The complexity is in maintaining a reasonable flex width balance between the prefix, number, and suffix elements. For example, in a tight context like the Border control:
> 
> <img alt="A BorderControl field that has a prefix, number field, and unit dropdown all in a 110px width" width="268" src="https://user-images.githubusercontent.com/555336/181244681-40c6ba1a-4b79-44de-b61b-e81c144016b8.png">
> 
> It might even be that we'll need to allow the consumer to customize the width via props, so it makes sense in different contexts.

mirka commented:
Follow-ups moved to #42879
